communicator/remote/command.go

```diff
@@ -59,23 +60,28 @@ func (c *Cmd) SetExitStatus(status int, err error) {
 	close(c.exitCh)
 }
 
-// Err returns any communicator related error.
-func (c *Cmd) Err() error {
+// Wait waits for the remote command to complete.
+// Wait may return an error from the communicator, or an ExitError if the
+// process exits with a non-zero exit status.
+func (c *Cmd) Wait() error {
+	<-c.exitCh
+
 	c.Lock()
 	defer c.Unlock()
 
-	return c.err
-}
+	if c.err != nil {
+		return c.err
+	}
 
-// ExitStatus returns the exit status of the remote command
-func (c *Cmd) ExitStatus() int {
-	c.Lock()
-	defer c.Unlock()
+	if c.exitStatus != 0 {
+		return ExitError(c.exitStatus)
+	}
 
-	return c.exitStatus
+	return nil
 }
 
-// Wait waits for the remote command to complete.
-func (c *Cmd) Wait() {
-	<-c.exitCh
+type ExitError int
+
+func (e ExitError) Error() string {
+	return fmt.Sprintf("exit status: %d", e)
```

Comment:
It seems like the most common thing to do with this error is to re-package it in a more verbose message that includes the command name. Should we just make the error type include that itself already so that this error has a useful error message by default? (We could still recognize it and replace it with a more specific error in cases where there's something more interesting to say, of course.)

It looks like we can get to the command as `c.Command` at the point where this error is generated, so should have everything we need to do it, at the expense of it needing to be a struct rather than just an integer.

Reply:
Nice idea. I was trying to keep it minimal, since users (though not in any cases here) often just want the integer exit code, and this matches the output of `*exec.ExitError`. Though since nothing in the provisioners as actually checking the code, maybe we wrap it up with the command string and get rid of all the `fmt.Errorf`s.